src/server/handler/mod.rs

```diff
@@ -1,60 +1,67 @@
 mod file_server;
+mod proxy;
 
 use anyhow::Result;
-use futures::Future;
+use async_trait::async_trait;
 use http::{Request, Response};
 use hyper::Body;
 use std::convert::TryFrom;
-use std::pin::Pin;
 use std::sync::Arc;
 use tokio::sync::Mutex;
 
 use crate::addon::file_server::FileServer;
+use crate::addon::proxy::Proxy;
 use crate::Config;
 
 use super::middleware::Middleware;
 
 use self::file_server::FileServerHandler;
+use self::proxy::ProxyHandler;
 
-/// The main handler for the HTTP request, a HTTP response is created
-/// as a result of this handler.
-///
-/// This handler will be executed against the HTTP request after every
-/// "Middleware Before" chain is executed but before any "Middleware After"
-/// chain is executed
-pub type Handler = Box<
-    dyn Fn(
-            Arc<Mutex<Request<Body>>>,
-        ) -> Pin<Box<dyn Future<Output = http::Response<Body>> + Send + Sync>>
-        + Send
-        + Sync,
->;
+/// A trait to implement on addons in order to handle incomming requests and
+/// generate responses.
+#[async_trait]
+pub trait RequestHandler {
+    async fn handle(&self, req: Arc<Mutex<Request<Body>>>) -> Arc<Mutex<http::Response<Body>>>;
+}
```

Comment:
The introduction of this async trait will open path to future addons to be implemented with ease.
In order to introduce main handlers now, the user will have to implement a single trait and provide it to Hyper as a service through the `main_service`.